src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -148,6 +148,7 @@ public class LockFileBuilder
                 }
             }
 
+            ValidateLockFileLibrary(lockFile.Libraries);
```

Comment:
Given that we're repairing the list after the fact, do we know why we got both here in the first place?

What's the list here when the project has a version that's different from the package.

Reply:
Thanks for the explanation.  

What do you think about removing this exception check and simply ranking the references?
Consider writing a generic method to avoid typos causing problems. 

We can chat offline if that's easier.

Reply:
@nkolev92 
Thank you for your suggestion. I revised the code, please check. I removed exception check since it's not possible scenario.
But writing generic method is little bit problematic. I can have something like below but problem is `second.Type` and `first.Type` part. Here `LockFileLibrary` and `LockFileTargetLibrary` both of them are already public api, so better not to make any changes to make them generic. Maybe I can use casting or reflection for that line but maybe it's overkill.

```
        private void ValidateLibrary<T, V>(IList<T> libraries, IGrouping<Tuple<string, Versioning.NuGetVersion>, V> item)
        {
            if (item.Count() == 2)
            {
                var first = item.First();
                var second = item.Last();

                // Prefer project reference over package reference, so remove the the package reference.
                libraries.Remove(RankReferences(second.Type) > RankReferences(first.Type) ? second : first);
            }
        }
```